components/browser/icons/src/main/java/mozilla/components/browser/icons/processor/ColorProcessor.kt

```diff
@@ -0,0 +1,87 @@
+import kotlin.math.max
+
+/**
+ * [IconProcessor] implementation to extract the dominant color from the icon.
+ */
+class ColorProcessor : IconProcessor {
+
+    override fun process(context: Context, request: IconRequest, resource: IconRequest.Resource?, icon: Icon): Icon {
+        if (icon.bitmap == null || icon.color != null) return icon
+
+        getEdgeColor(icon.bitmap)?.let { edgeColor ->
+            return icon.copy(color = edgeColor)
+        }
+
+        val dominantColor = Palette.from(icon.bitmap).generate().getDominantColor(DEFAULT_COLOR)
+        return icon.copy(color = dominantColor and DOMINANT_COLOR_MASK)
+    }
+
+    companion object {
+        /**
+         * No color, here we use photon color
+         */
+        private val DEFAULT_COLOR = rgb(177, 177, 179)
+
+        private const val DOMINANT_COLOR_MASK = 0x7FFFFFFF
+
+        /**
+         * If a bitmap has a consistent edge colour (i.e. if all the border pixels have the same colour),
+         * return that colour. Return null otherwise.
+         */
+        @Suppress("ReturnCount")
+        @ColorInt
+        private fun getEdgeColor(bitmap: Bitmap): Int? {
+            val width = bitmap.width
+            val height = bitmap.height
+            // Only allocate an array once, with the max width we need once, to minimise the number
+            // of allocations.
+            val edge = IntArray(max(width, height))
+
+            // Top:
+            bitmap.getPixels(edge, 0, width, 0, 0, width, 1)
+            @ColorInt val edgeColor = getEdgeColorFromSingleDimension(edge, width) ?: return null
+
+            // Bottom:
+            bitmap.getPixels(edge, 0, width, 0, height - 1, width, 1)
+            if (getEdgeColorFromSingleDimension(edge, width) != edgeColor) return null
+
+            // Left:
+            bitmap.getPixels(edge, 0, 1, 0, 0, 1, height)
+            if (getEdgeColorFromSingleDimension(edge, height) != edgeColor) return null
+
+            // Right:
+            bitmap.getPixels(edge, 0, 1, width - 1, 0, 1, height)
+            if (getEdgeColorFromSingleDimension(edge, height) != edgeColor) return null
+
+            return edgeColor
+        }
+
+        /**
+         * Obtain the colour for a given edge if all colors are the same.
+         *
+         * @param edge An array containing the color values of the pixels constituting the edge of a bitmap.
+         * @param length The length of the array to be traversed. Must be smaller than, or equal to
+         * the total length of the array.
+         * @return The colour contained within the array, or null if colours vary.
+         */
+        @ColorInt
+        private fun getEdgeColorFromSingleDimension(edge: IntArray, length: Int): Int? {
+            @ColorInt val color = edge[0]
+            val allSameColor = edge.asSequence().drop(1).take(length - 1).all { it == color }
+            return if (allSameColor) color else null
+        }
```

Comment:
Nit: put these `private fun` functions into the processor itself, rather than the companion object.